core/trino-main/src/main/java/io/trino/metadata/InternalFunctionBundle.java

```diff
@@ -138,8 +138,9 @@ public ScalarFunctionImplementation getScalarFunctionImplementation(
 
     private SpecializedSqlScalarFunction specializeScalarFunction(FunctionId functionId, BoundSignature boundSignature, FunctionDependencies functionDependencies)
     {
-        SqlScalarFunction function = (SqlScalarFunction) getSqlFunction(functionId);
-        return function.specialize(boundSignature, functionDependencies);
+        SqlFunction function = getSqlFunction(functionId);
+        checkArgument(function instanceof SqlScalarFunction, "%s is not a scalar function", function.getFunctionMetadata().getSignature());
```

Comment:
We need unit tests for that and this is an SQL error so we need a proper UX error.

Reply:
On my eye it is not SQL error. It is kind of invariant that should be satisfied here. We have similar here: 
https://github.com/trinodb/trino/commit/ad637f702ef272f4b96eca006fb047379f4d0604#diff-bd8fefc1e4638b7f65e377b75c16b46a2c541a24a64993860663507351917230R106
Are sure that we should test every invariant? This invariant is to help detect such problem immediately in the future. But, the end user should observe this (if there is no bug in code).

Reply:
This should already be handled during SQL analysis by `io.trino.sql.analyzer.Analyzer#verifyNoAggregateWindowOrGroupingFunctions` as the newly added test shows.

Reply:
Yes, this case is handled by `StatementAnalyzer` now. This is why I've put assertions, as the last ressort. [The first PR did not include `verifyNoAggregateWindowOrGroupingFunctions` so it was true then that it was exposed to the end user].